File: crates/pulsevm_ffi/builtins.cpp
#include <cstring>

#include <compiler_builtins.hpp>
#include <softfloat.hpp>

#include <fc/uint128.hpp>

#include <pulsevm_ffi/src/bridge.rs.h>

namespace pulsevm::chain {

inline unsigned __int128 to_u128(const U128& v) {
    return (static_cast<unsigned __int128>(v.hi) << 64)
         |  static_cast<unsigned __int128>(v.lo);
}

inline U128 from_u128(unsigned __int128 x) {
    return U128{
        static_cast<uint64_t>(x),          // lo
        static_cast<uint64_t>(x >> 64),    // hi
    };
}

inline __int128 to_i128(const I128& v) {
    return (static_cast<__int128>(static_cast<unsigned __int128>(v.hi) << 64)
          | static_cast<__int128>(static_cast<unsigned __int128>(v.lo)));
}

inline I128 from_i128(__int128 x) {
    return I128{
        static_cast<uint64_t>(x),                              // lo
        static_cast<uint64_t>(static_cast<unsigned __int128>(x) >> 64), // hi
    };
}

// arithmetic long double
Float128 addtf3( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    float128_t a = {{ la, ha }};
    float128_t b = {{ lb, hb }};
    float128_t result = f128_add( a, b );
    return Float128{ result.v[0], result.v[1] };
}
Float128 subtf3( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    float128_t a = {{ la, ha }};
    float128_t b = {{ lb, hb }};
    float128_t result = f128_sub( a, b );
    return Float128{ result.v[0], result.v[1] };
}
Float128 multf3( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    float128_t a = {{ la, ha }};
    float128_t b = {{ lb, hb }};
    float128_t result = f128_mul( a, b );
    return Float128{ result.v[0], result.v[1] };
}
Float128 divtf3( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    float128_t a = {{ la, ha }};
    float128_t b = {{ lb, hb }};
    float128_t result = f128_div( a, b );
    return Float128{ result.v[0], result.v[1] };
}
Float128 negtf2( uint64_t la, uint64_t ha ) {
    float128_t result = {{ la, (ha ^ (uint64_t)1 << 63) }};
    return Float128{ result.v[0], result.v[1] };
}

// conversion long double
Float128 extendsftf2( float f ) {
    float128_t result = f32_to_f128( to_softfloat32(f) );
    return Float128{ result.v[0], result.v[1] };
}
Float128 extenddftf2( double d ) {
    float128_t result = f64_to_f128( to_softfloat64(d) );
    return Float128{ result.v[0], result.v[1] };
}
double trunctfdf2( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    return from_softfloat64(f128_to_f64( f ));
}
float trunctfsf2( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    return from_softfloat32(f128_to_f32( f ));
}
int32_t fixtfsi( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    return f128_to_i32( f, 0, false );
}
int64_t fixtfdi( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    return f128_to_i64( f, 0, false );
}
I128 fixtfti( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    __int128_t result = ___fixtfti( f );
    return from_i128(result);
}
uint32_t fixunstfsi( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    return f128_to_ui32( f, 0, false );
}
uint64_t fixunstfdi( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    return f128_to_ui64( f, 0, false );
}
U128 fixunstfti( uint64_t l, uint64_t h ) {
    float128_t f = {{ l, h }};
    auto result = ___fixunstfti( f );
    return from_u128(result);
}
I128 fixsfti( float a ) {
    __int128_t result = ___fixsfti( to_softfloat32(a).v );
    return from_i128(result);
}
I128 fixdfti( double a ) {
    __int128_t result = ___fixdfti( to_softfloat64(a).v );
    return from_i128(result);
}
U128 fixunssfti( float a ) {
    auto result = ___fixunssfti( to_softfloat32(a).v );
    return from_u128(result);
}
U128 fixunsdfti( double a ) {
    auto result = ___fixunsdfti( to_softfloat64(a).v );
    return from_u128(result);
}
double floatsidf( int32_t i ) {
    return from_softfloat64(i32_to_f64(i));
}
Float128 floatsitf( int32_t i ) {
    float128_t result = i32_to_f128(i);
    return Float128{ result.v[0], result.v[1] };
}
Float128 floatditf( uint64_t a ) {
    float128_t result = i64_to_f128( a );
    return Float128{ result.v[0], result.v[1] };
}
Float128 floatunsitf( uint32_t i ) {
    float128_t result = ui32_to_f128(i);
    return Float128{ result.v[0], result.v[1] };
}
Float128 floatunditf( uint64_t a ) {
    float128_t result = ui64_to_f128( a );
    return Float128{ result.v[0], result.v[1] };
}
double floattidf( uint64_t l, uint64_t h ) {
    fc::uint128 v(h, l);
    unsigned __int128 val = (unsigned __int128)v;
    return ___floattidf( *(__int128*)&val );
}
double floatuntidf( uint64_t l, uint64_t h ) {
    fc::uint128 v(h, l);
    return ___floatuntidf( (unsigned __int128)v );
}

int unordtf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    float128_t a = {{ la, ha }};
    float128_t b = {{ lb, hb }};
    if ( f128_is_nan(a) || f128_is_nan(b) )
        return 1;
    return 0;
}

inline static int cmptf2_impl( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb, int return_value_if_nan ) {
    float128_t a = {{ la, ha }};
    float128_t b = {{ lb, hb }};
    if ( unordtf2(la, ha, lb, hb) )
        return return_value_if_nan;
    if ( f128_lt( a, b ) )
        return -1;
    if ( f128_eq( a, b ) )
        return 0;
    return 1;
}
int eqtf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, 1);
}
int netf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, 1);
}
int getf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, -1);
}
int gttf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, 0);
}
int letf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, 1);
}
int lttf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, 0);
}
int cmptf2( uint64_t la, uint64_t ha, uint64_t lb, uint64_t hb ) {
    return cmptf2_impl(la, ha, lb, hb, 1);
}

}

File: crates/pulsevm_core/src/chain/webassembly/tests/builtins.rs
use super::*;

type Binary =
    fn(FunctionEnvMut<WasmContext>, WasmPtr<u8>, u64, u64, u64, u64) -> Result<(), RuntimeError>;
type Shift =
    fn(FunctionEnvMut<WasmContext>, WasmPtr<u8>, u64, u64, u32) -> Result<(), RuntimeError>;
type Compare = fn(FunctionEnvMut<WasmContext>, u64, u64, u64, u64) -> Result<i32, RuntimeError>;

#[test]
fn integer_shifts_cover_limb_and_width_boundaries() {
    let mut h = Host::new(false);
    for (op, arithmetic, left) in [
        (__ashlti3 as Shift, false, true),
        (__lshlti3, false, true),
        (__lshrti3, false, false),
        (__ashrti3, true, false),
    ] {
        for value in [0, 1, (1u128 << 127) | 0x123456789abcdef, u128::MAX] {
            for shift in [0, 1, 63, 64, 65, 127, 128, 129, u32::MAX] {
                op(h.env(), ptr(OUT), value as u64, (value >> 64) as u64, shift).unwrap();
                let expected = if arithmetic {
                    ((value as i128) >> (shift % 128)) as u128
                } else if shift >= 128 {
                    0
                } else if left {
                    value << shift
                } else {
                    value >> shift
                };
                assert_eq!(h.read128(), expected, "value={value:x}, shift={shift}");
            }
        }
        assert!(op(h.env(), ptr(END - 15), 1, 0, 0).is_err());
        h.budget(cost::BUILTIN - 1);
        assert!(op(h.env(), ptr(OUT), 1, 0, 0).is_err());
        assert_eq!(h.remaining(), 0);
        h.budget(u64::MAX);
    }
}

#[test]
fn integer_arithmetic_handles_signs_overflow_and_zero_divisors() {
    let mut h = Host::new(false);
    for (op, lhs, rhs, expected, price) in [
        (
            __divti3 as Binary,
            (-17i128) as u128,
            5,
            (-3i128) as u128,
            cost::BUILTIN_DIV,
        ),
        (
            __modti3,
            (-17i128) as u128,
            5,
            (-2i128) as u128,
            cost::BUILTIN_DIV,
        ),
        (
            __divti3,
            i128::MIN as u128,
            u128::MAX,
            i128::MIN as u128,
            cost::BUILTIN_DIV,
        ),
        (__modti3, i128::MIN as u128, u128::MAX, 0, cost::BUILTIN_DIV),
        (__udivti3, u128::MAX, 2, u128::MAX / 2, cost::BUILTIN_DIV),
        (__umodti3, u128::MAX, 2, 1, cost::BUILTIN_DIV),
        (__multi3, u128::MAX, 2, u128::MAX - 1, cost::BUILTIN),
    ] {
        h.budget(price);
        op(
            h.env(),
            ptr(OUT),
            lhs as u64,
            (lhs >> 64) as u64,
            rhs as u64,
            (rhs >> 64) as u64,
        )
        .unwrap();
        assert_eq!(h.read128(), expected);
        assert_eq!(h.remaining(), 0);
        h.budget(u64::MAX);
        assert!(op(h.env(), ptr(END - 15), 9, 0, 2, 0).is_err());
    }
    for op in [__divti3 as Binary, __udivti3, __modti3, __umodti3] {
        h.write(OUT, &[0xa5; 16]);
        assert_error(op(h.env(), ptr(OUT), 1, 0, 0, 0), "zero");
        assert_eq!(h.read(OUT, 16), vec![0xa5; 16]);
    }
}

// Explicit IEEE binary128 encodings, independent of the conversion helpers under test.
const ONE: u128 = 0x3fff0000000000000000000000000000;
const TWO: u128 = 0x40000000000000000000000000000000;
const THREE: u128 = 0x40008000000000000000000000000000;
const HALF: u128 = 0x3ffe0000000000000000000000000000;
const SIGN: u128 = 1 << 127;

#[test]
fn binary128_arithmetic_writes_exact_unaligned_results() {
    let mut h = Host::new(false);
    for (op, a, b, expected, price) in [
        (__addtf3 as Binary, ONE, TWO, THREE, cost::BUILTIN),
        (__subtf3, THREE, TWO, ONE, cost::BUILTIN),
        (__multf3, HALF, TWO, ONE, cost::BUILTIN),
        (__divtf3, ONE, TWO, HALF, cost::BUILTIN_DIV),
    ] {
        h.budget(price);
        op(
            h.env(),
            ptr(OUT),
            a as u64,
            (a >> 64) as u64,
            b as u64,
            (b >> 64) as u64,
        )
        .unwrap();
        assert_eq!(h.read128(), expected);
        assert_eq!(h.remaining(), 0);
        h.budget(u64::MAX);
        assert!(op(h.env(), ptr(END - 15), 0, 0, 0, 0).is_err());
    }
    __negtf2(h.env(), ptr(OUT), 0, (ONE >> 64) as u64).unwrap();
    assert_eq!(h.read128(), ONE | SIGN);
    __negtf2(h.env(), ptr(OUT), 0, 0).unwrap();
    assert_eq!(h.read128(), SIGN);
}

#[test]
fn binary128_conversions_preserve_sign_and_truncate_toward_zero() {
    let mut h = Host::new(false);
    __extendsftf2(h.env(), ptr(OUT), -1.0).unwrap();
    assert_eq!(h.read128(), ONE | SIGN);
    __extenddftf2(h.env(), ptr(OUT), 0.5).unwrap();
    assert_eq!(h.read128(), HALF);
    let hi = ((THREE | SIGN) >> 64) as u64;
    assert_eq!(__trunctfdf2(h.env(), 0, hi).unwrap(), -3.0);
    assert_eq!(__trunctfsf2(h.env(), 0, hi).unwrap(), -3.0);
    assert_eq!(__fixtfsi(h.env(), 0, hi).unwrap(), -3);
    assert_eq!(__fixtfdi(h.env(), 0, hi).unwrap(), -3);
    __fixtfti(h.env(), ptr(OUT), 0, hi).unwrap();
    assert_eq!(h.read128() as i128, -3);
    let hi = (THREE >> 64) as u64;
    assert_eq!(__fixunstfsi(h.env(), 0, hi).unwrap(), 3);
    assert_eq!(__fixunstfdi(h.env(), 0, hi).unwrap(), 3);
    __fixunstfti(h.env(), ptr(OUT), 0, hi).unwrap();
    assert_eq!(h.read128(), 3);
    __fixsfti(h.env(), ptr(OUT), -3.75).unwrap();
    assert_eq!(h.read128() as i128, -3);
    __fixdfti(h.env(), ptr(OUT), -3.75).unwrap();
    assert_eq!(h.read128() as i128, -3);
    __fixunssfti(h.env(), ptr(OUT), 3.75).unwrap();
    assert_eq!(h.read128(), 3);
    __fixunsdfti(h.env(), ptr(OUT), 3.75).unwrap();
    assert_eq!(h.read128(), 3);
    assert_eq!(__floatsidf(h.env(), -3).unwrap(), -3.0);
    __floatsitf(h.env(), ptr(OUT), -3).unwrap();
    assert_eq!(h.read128(), THREE | SIGN);
    __floatditf(h.env(), ptr(OUT), (-3i64) as u64).unwrap();
    assert_eq!(h.read128(), THREE | SIGN);
    __floatunsitf(h.env(), ptr(OUT), 3).unwrap();
    assert_eq!(h.read128(), THREE);
    __floatunditf(h.env(), ptr(OUT), 3).unwrap();
    assert_eq!(h.read128(), THREE);
    assert_eq!(
        __floattidf(h.env(), (-3i64) as u64, u64::MAX).unwrap(),
        -3.0
    );
    assert_eq!(__floatuntidf(h.env(), 3, 0).unwrap(), 3.0);
}

#[test]
fn binary128_comparisons_distinguish_order_equality_and_nan() {
    let mut h = Host::new(false);
    for (op, unordered) in [
        (__eqtf2 as Compare, 1),
        (__netf2, 1),
        (__getf2, -1),
        (__gttf2, 0),
        (__letf2, 1),
        (__lttf2, 0),
        (__cmptf2, 1),
    ] {
        for (a, b, expected) in [
            (ONE, TWO, -1),
            (TWO, ONE, 1),
            (ONE, ONE, 0),
            (0, SIGN, 0),
            (0x7fff8000000000000000000000000000, ONE, unordered),
        ] {
            assert_eq!(
                op(
                    h.env(),
                    a as u64,
                    (a >> 64) as u64,
                    b as u64,
                    (b >> 64) as u64
                )
                .unwrap(),
                expected
            );
        }
    }
    assert_eq!(
        __unordtf2(h.env(), 0, 0x7fff800000000000, 0, (ONE >> 64) as u64).unwrap(),
        1
    );
    assert_eq!(__unordtf2(h.env(), 0, 0, 0, 0).unwrap(), 0);
}
